src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/ConnectionPool/ChannelDbConnectionPool.cs

```diff
@@ -111,9 +228,383 @@ public void TransactionEnded(Transaction transaction, DbConnectionInternal trans
+                // to GetInternalConnection, which means it will not use Task.Run or any async-await logic that would
+                // schedule tasks on the managed threadpool.
+                connection = task.ConfigureAwait(false).GetAwaiter().GetResult();
+                return connection is not null;
+            }
+
+            // Early exit if the task is already completed.
+            if (taskCompletionSource.Task.IsCompleted)
+            {
+                connection = null;
+                return false;
+            }
+
+            /* 
+                * This is ugly, but async anti-patterns above and below us in the stack necessitate a fresh task to be
+                * created. Ideally we would just return the Task from GetInternalConnection and let the caller await 
+                * it as needed, but instead we need to signal to the provided TaskCompletionSource when the connection
+                * is established. This pattern has implications for connection open retry logic that are intricate 
+                * enough to merit dedicated work. For now, callers that need to open many connections asynchronously 
+                * and in parallel *must* pre-prevision threads in the managed thread pool to avoid exhaustion and 
+                * timeouts.
+                * 
+                * Also note that we don't have access to the cancellation token passed by the caller to the original 
+                * OpenAsync call. This means that we cannot cancel the connection open operation if the caller's token
+                * is cancelled. We can only cancel based on our own timeout, which is set to the owningObject's 
+                * ConnectionTimeout. 
+                */
+            Task.Run(async () =>
+            {
+                if (taskCompletionSource.Task.IsCompleted)
+                {
+                    return;
+                }
+
+                // We're potentially on a new thread, so we need to properly set the ambient transaction.
+                // We rely on the caller to capture the ambient transaction in the TaskCompletionSource's AsyncState
+                // so that we can access it here. Read: area for improvement.
+                ADP.SetCurrentTransaction(taskCompletionSource.Task.AsyncState as Transaction);
+
+                try
+                {
+                    var connection = await GetInternalConnection(
+                        owningObject, 
+                        userOptions, 
+                        TimeSpan.FromSeconds(owningObject.ConnectionTimeout), 
+                        async: true, 
+                        CancellationToken.None
+                    ).ConfigureAwait(false);
+                    taskCompletionSource.SetResult(connection);
+                }
+                catch (Exception e)
+                {
+                    taskCompletionSource.SetException(e);
+                }
```

Comment:
If I understand the code correctly, `taskCompletionSource` passed here is the one from `SqlConnection.InternalOpenAsync` (specifically `completion`), which can be cancelled via `CancellationToken` passed there. That means that it's possible to get an exception from `taskCompletionSource.SetResult` and `taskCompletionSource.SetException`, resulting in lost connections.